crates/invoker-impl/src/lib.rs

```diff
@@ -371,12 +425,18 @@ where
                     }
                 }
             },
-
-            Some(invoke_input_command) = segmented_input_queue.dequeue(), if !segmented_input_queue.is_empty() && self.quota.is_slot_available() => {
+            item = action_token_stream.next(), if self.action_token.is_none() => {
+                self.action_token = item;
+            },
+            item = invocation_token_stream.next(), if self.invocation_token.is_none() => {
+                self.invocation_token = item;
+            },
+            Some(invoke_input_command) = segmented_input_queue.dequeue(), if self.invocation_token.is_some() && !segmented_input_queue.is_empty() && self.quota.is_slot_available() => {
+                self.invocation_token.take();
                 self.handle_invoke(options, invoke_input_command.partition, invoke_input_command.invocation_id, invoke_input_command.invocation_epoch, invoke_input_command.invocation_target, invoke_input_command.journal);
             },
-
-            Some(invocation_task_msg) = self.invocation_tasks_rx.recv() => {
+            Some(invocation_task_msg) = self.invocation_tasks_rx.recv(), if self.action_token.is_some() => {
+                self.action_token.take();
```

Comment:
I am wondering if not all actions should "consume" the token. A NewEntry or Command will most probably consume the token. Other actions like (pinned, close, suspend or fail) shouldn't.